meteor_packages/mats-common/public/python/mode_stats.py

```diff
@@ -0,0 +1,313 @@
+            n_o = 0
+            for key in mode_header_lookup:
+                n_f = n_f + mode_header_lookup[key]["lookup_f_index"]
+                n_o = n_o + mode_header_lookup[key]["lookup_o_index"]
+            ofb = n_f / n_o
+        else:
+            ofb = 'null'
+    except TypeError as e:
+        error = "Error calculating ofb: " + str(e)
+        ofb = 'null'
+    except ValueError as e:
+        error = "Error calculating ofb: " + str(e)
+        ofb = 'null'
+    return ofb
+
+
+# function for calculating mean centroid distance from MET MODE output
+def calculate_mcd(sub_interest, sub_pair_fid, sub_pair_oid, sub_mode_header_id, sub_cent_dist):
+    global error
+    total_error_array = []
+    try:
+        if len(sub_pair_fid) > 0 and len(sub_pair_oid) > 0:
+            # Mean distance calculated in a general sense using the same method as for OTS
+            # this method does not require object "matches"
+            # Sort the pair_interest array but keep that sorted interest linked with the pair object IDs
+            indices = np.argsort(sub_interest)
+            # reverse the indices array so that it goes descending from maximum
+            indices = indices[::-1]
+            sorted_fid = sub_pair_fid[indices]
+            sorted_oid = sub_pair_oid[indices]
+            sorted_mode_header = sub_mode_header_id[indices]
+            sorted_cent_dist = sub_cent_dist[indices]
+
+            matched_fid = {}
+            matched_oid = {}
+            for i in range(0, len(sub_interest)):
+                this_mode_header_id = str(sorted_mode_header[i])
+                this_fid = sorted_fid[i]
+                this_oid = sorted_oid[i]
+                this_cent_dist = sorted_cent_dist[i]
+                # f_lat = individual_obj_lookup[this_mode_header_id][this_fid]["centroid_lat"]
+                # o_lat = individual_obj_lookup[this_mode_header_id][this_oid]["centroid_lat"]
+                # f_lon = individual_obj_lookup[this_mode_header_id][this_fid]["centroid_lon"]
+                # o_lon = individual_obj_lookup[this_mode_header_id][this_oid]["centroid_lon"]
+                if this_mode_header_id not in matched_fid.keys():
+                    matched_fid[this_mode_header_id] = []
+                    matched_oid[this_mode_header_id] = []
+                if (this_fid not in matched_fid[this_mode_header_id]) and (this_oid not in matched_oid[this_mode_header_id]):
+                    matched_fid[this_mode_header_id].append(this_fid)
+                    matched_oid[this_mode_header_id].append(this_oid)
+                    # # Compute E-W and N-S error components for generalized matches
+                    # x_error = np.sign(f_lon - o_lon) * gc_dist(f_lon, o_lat, o_lon, o_lat)
+                    # y_error = np.sign(f_lat - o_lat) * gc_dist(o_lon, f_lat, o_lon, o_lat)
+                    # total_error = (x_error ** 2 + y_error ** 2) ** 0.5
+                    # total_error_array.append(total_error)
```

Comment:
Same as in my last comment, I think that there's a 30% chance Jeff Duda will change his mind and we will use that code, so I've left it in for the time being. I didn't want to just rely on the git history because it went through a few versions and this was the correct one. If he hasn't decided that we want it in two more METexpress versions then I will remove it.